File: Week 6/C++ Solutions/treeTraversals.cpp
#include <iostream>
#include <vector>

using std::cin;
using std::cout;
using std::vector;

class TreeOrder {
public:
  long int numNodes;
  vector<int> keys;
  vector<int> lefts;
  vector<int> rights;

  void read() {
    int k, l, r;

    cin >> numNodes;
    keys.resize(numNodes);
    lefts.resize(numNodes);
    rights.resize(numNodes);

    for (int i = 0; i < numNodes; i++) {
      cin >> k >> l >> r;
      keys[i] = k; lefts[i] = l; rights[i] = r;
    }
  }

  void inOrderTraversal(int root = 0) {
    if (root == -1) {return;}
    inOrderTraversal(this->lefts[root]);
    cout << this->keys[root] << " ";
    inOrderTraversal(this->rights[root]);
  }

  void preOrderTraversal(int root = 0) {
    if (root == -1) {return;}
    cout << this->keys[root] << " ";
    preOrderTraversal(this->lefts[root]);
    preOrderTraversal(this->rights[root]);
  }

  void postOrderTraversal(int root = 0) {
    if (root == -1) {return;}
    postOrderTraversal(this->lefts[root]);
    postOrderTraversal(this->rights[root]);
    cout << this->keys[root] << " ";
  }
};

int main() {
  TreeOrder tree;
  tree.read();
  tree.inOrderTraversal();
  cout << "\n";
  tree.preOrderTraversal();
  cout << "\n";
  tree.postOrderTraversal();
  return 0;
}

File: Week 6/treeTraversals.rb
class TreeOrder
  def read()
    @numNodes = gets.chomp.to_i
    @keys = Array.new(@numNodes, 0)
    @lefts = Array.new(@numNodes, 0)
    @rights = Array.new(@numNodes, 0)

    (0...@numNodes).each do |i|
      k, l, r = gets.chomp.split(/ /).map(&:to_i)
      @keys[i] = k ; @lefts[i] = l ; @rights[i] = r
    end
  end

  def inOrderTraversal(root = 0)
    return if root == -1
    self.inOrderTraversal(@lefts[root])
    print "#{self.keys[root]} "
    self.inOrderTraversal(@rights[root])
  end

  def preOrderTraversal(root = 0)
    return if root == -1
    print "#{@keys[root]} "
    self.preOrderTraversal(@lefts[root])
    self.preOrderTraversal(@rights[root])
  end

  def postOrderTraversal(root = 0)
    return if root == -1
    self.postOrderTraversal(@lefts[root])
    self.postOrderTraversal(@rights[root])
    print "#{@keys[root]} "
  end
end

tree = TreeOrder.new()
tree.read()
tree.inOrderTraversal()
puts ""
tree.preOrderTraversal()
puts ""
tree.postOrderTraversal()
